Bio/SeqUtils/NucUtils.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2018 Aleksandra Jarmolinska.  All rights reserved.
+# This file is part of the Biopython distribution and governed by your
+# choice of the "Biopython License Agreement" or the "BSD 3-Clause License".
+# Please see the LICENSE file that should have been included as part of this
+# package.
+"""Functions for dealing with GC/AT content."""
+
+
+from __future__ import print_function
+
+__GC_CONTENT_VALUE = {
+    "G": 1.,
+    "C": 1.,
+    "A": 0.,
+    "T": 0.,
+    "S": 1.,
+    "W": 0.,
+    "N": 0.5
+}
+__GC_CONTENT_VALUE_AMBG = {
+    "M": 0.5,
+    "K": 0.5,
+    "R": 0.5,
+    "Y": 0.5,
+
+    "B": 0.66,
+    "D": 0.33,
+    "H": 0.33,
+    "V": 0.66,
+}
+
+
+def calc_gc_content(seq, interpret_all=False):
+    """Calculate G+C content, returns a fraction (float between 0 and 1).
+
+    Copes mixed case sequences, and with the ambiguous nucleotides S (G or C)
+    and N ( A or C or G or T ) when counting the G and C content.
+
+    When interpret_all is True, includes probability of other ambiguous
+    to be C or G:
+    - for M,K,R,Y it is 0.5
+    - for B,V it is 0.66
+    - for D,H it is 0.33.
+
+    The percentage is calculated against the full length, e.g.:
+
+    >>> from Bio.SeqUtils.NucUtils import calc_gc_content
+    >>> calc_gc_content("ACTGS")
+    0.6
+
+    >>> from Bio.SeqUtils.NucUtils import calc_gc_content
+    >>> calc_gc_content("ACTGN")
+    0.5
+
+    >>> from Bio.SeqUtils.NucUtils import calc_gc_content
+    >>> calc_gc_content("ACTGB")
+    0.4
+
+    >>> from Bio.SeqUtils.NucUtils import calc_gc_content
+    >>> calc_gc_content("ACTGB",True)
+    0.532
+
+    Note that this will return zero for an empty sequence.
+    """
+    len_seq = len(seq)
+    if not len_seq:
+        return 0.
+
+    gc = sum(__GC_CONTENT_VALUE.get(x.upper(),
```

Comment:
I worry that calling ``x.upper()`` for every letter in a large sequence like an entire chromosome would be far slower than the existing approach in ``Bio/SeqUtils/__init__.py`` which has both cases in the dictionary. Have you benchmarked this?

Reply:
You are right - I benchmarked it just now, and it increases the times approx. by half. I'm now adding the probabilities calculation based on IUPAC tables, and will correct the dictionaries accordingly.